src/cursor.c

```diff
@@ -307,11 +265,86 @@ process_cursor_motion_out_of_surface(struct server *server, uint32_t time)
 	wlr_seat_pointer_notify_motion(server->seat.seat, time, sx, sy);
 }
 
+/*
+ * Common logic shared by cursor_update_focus() and process_cursor_motion()
+ */
 static void
-process_cursor_motion(struct server *server, uint32_t time)
+cursor_update_common(struct server *server, struct wlr_scene_node *node,
+		struct wlr_surface *surface, double sx, double sy,
+		enum ssd_part_type view_area, uint32_t time_msec,
+		bool cursor_has_moved)
 {
-	static bool cursor_name_set_by_server;
+	struct seat *seat = &server->seat;
+	struct wlr_seat *wlr_seat = seat->seat;
 
+	ssd_update_button_hover(node, &server->ssd_hover_state);
+
+	if (server->input_mode != LAB_INPUT_STATE_PASSTHROUGH) {
+		/*
+		 * Prevent updating focus/cursor image during
+		 * interactive move/resize
+		 */
+		return;
+	}
+
+	/* TODO: verify drag_icon logic */
+	if (seat->pressed.surface && surface != seat->pressed.surface
+			&& !seat->drag_icon) {
+		if (cursor_has_moved) {
+			/*
+			 * Button has been pressed while over another
+			 * surface and is still held down.  Just send
+			 * the motion events to the focused surface so
+			 * we can keep scrolling or selecting text even
+			 * if the cursor moves outside of the surface.
+			 */
+			process_cursor_motion_out_of_surface(server, time_msec);
+		}
+		return;
+	}
+
+	if (surface && !input_inhibit_blocks_surface(seat, surface->resource)) {
+		/*
+		 * Cursor is over an input-enabled client surface.  The
+		 * cursor image will be set by request_cursor_notify()
+		 * in response to the enter event.
+		 */
+		if (surface != wlr_seat->pointer_state.focused_surface
+				|| !seat->client_cursor_requested) {
+			/*
+			 * Enter the surface if necessary.  Usually we
+			 * prevent re-entering an already focused
+			 * surface, because the extra leave and enter
+			 * events can confuse clients (e.g. break
+			 * double-click detection).
+			 *
+			 * We do however send a leave/enter event pair
+			 * if we need to trigger a cursor image update
+			 * (!client_cursor_requested).
+			 */
+			wlr_seat_pointer_notify_clear_focus(wlr_seat);
+			wlr_seat_pointer_notify_enter(wlr_seat, surface,
+				sx, sy);
+			seat->client_cursor_requested = true;
+		}
+		if (cursor_has_moved) {
+			wlr_seat_pointer_notify_motion(wlr_seat, time_msec,
+				sx, sy);
+		}
+	} else {
+		/*
+		 * Cursor is over a server (labwc) surface.  Clear focus
+		 * from the focused client (if any, no-op otherwise) and
+		 * set the cursor image ourselves.
+		 */
+		wlr_seat_pointer_notify_clear_focus(wlr_seat);
+		set_server_cursor(seat, view_area);
```

Comment:
If a surface was blocked due to `input_inhibit` this will clear the focus which might or might not be what we should do.
It is unrelated to the PR though as it was the same before AFAIR. We should still verify what should happen in that case as some follow up.

Reply:
#538 might (or might not) be related.